keps/sig-auth/20190116-service-account-external-signing.md

```diff
@@ -0,0 +1,152 @@
+  * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+The Kubernetes API server has always read service account keys from disk as the process starts, and kept them in memory for the duraiton of the server's lifetime. As the API server can now verify and issue projected volume tokens, it would be advantageous to support external signing and verifying of token data over an API, as well as reading public keys from an API.
+
+## Motivation
+
+For operators who want to regularly rotate the signing and verifying keys for projected volume tokens, the Kubernetes API server must be restarted in order to use a new key. To facilitate easy key rotation, this KEP includes an proposal for a grpc API to support out of process signing and listing of signing keys.
+
+### Goals
+
+- Support for out-of-process JWT signing
+- Support for listing public verifying keys
+- Preserve existing behavior and performance for keys not read over a socket
+
+### Non-Goals
+
+- Reading TLS serving certificates and key from a socket or reloading of the API server with new cert and key
+- Reading any other certificates from a file
+
+## Proposal
+
+### Preserve existing behavior
+
+The API server flags `--service-account-key-file` and `--service-account-signing-key-file` will continue be used for reading from files. The API server will detect if the provided files are regular files to be read once at process startup, or unix sockets that should be read on each token request using a new API. Legacy service account verification will skip any socket files.
+
+### Updates to API server token generation
+
+As of Kubernetes v1.13.2, the API server uses the functions `JWTTokenGenerator` and `JWTTokenAuthenticator`. New types that implement the `TokenGenerator` interface and support token validation will be added to `k8s.io/kubernetes/pkg/serviceaccount/`.
+
+### New API
+
+I'm proposing creating a new versioned grpc API under `k8s.io/kubernetes/pkg/serviceaccount`. This will be similar to how the KMS envelope encryption has an API at `k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/service.proto`
+
+```proto
+syntax = "proto3";
+
+package v1beta1;
+
+service KeyService {
+  // Sign an incoming payload
+  rpc SignPayload(SignPayloadRequest) returns (SignPayloadResponse) {}
+  // List supported Algorithms
+  rpc ListAlgorithms(ListAlgorithmsRequest) returns (ListAlgorithmsResponse) {}
+  // Get a specific key by ID, defaulting to the current public key
+  rpc GetPublicKey(GetPublicKeyRequest) returns (GetPublicKeyResponse) {}
+  // List all active public keys
+  rpc ListPublicKeys(ListPublicKeysRequest) returns (ListPublicKeysResponse) {}
+}
+
+message SignPayloadRequest {
+  // payload is the content to be signed
+  bytes payload = 1;
+  // algorithm specifies which algorithm to sign with
+  string algorithm = 2;
```

Comment:
I'm not sure if this is useful. The apiserver reads this from the ListPublicKeys response than passes it back to the signer? What happens if it gets it wrong e.g. if a key is rotated to a new algorithm?

Reply:
I see that we are somewhat constrained by the OpaqueSigner interface.

Reply:
I agree with the algorithm comment, it should be a list. A list would allow to specify algorithms and you are able to migrate, in case one algorithm is broken.

Reply:
We are limited in what we can do here because `signer.Public().Algorithm` is the `alg` value for `SignPayload`.  I am trying to think if there is some clever way we can allow moving to a new algorithm for new signed payloads while verifying previously signed data with the old algorithm.

We need something like an `OpaqueVerifier`...

```go
// OpaqueSigner is an interface that supports signing payloads with opaque
// private key(s). Private key operations preformed by implementors may, for
// example, occur in a hardware module. An OpaqueSigner may rotate signing keys
// transparently to the user of this interface.
type OpaqueSigner interface {
	// Public returns the public key of the current signing key.
	Public() *JSONWebKey
	// Algs returns a list of supported signing algorithms.
	Algs() []SignatureAlgorithm
	// SignPayload signs a payload with the current signing key using the given
	// algorithm.
	SignPayload(payload []byte, alg SignatureAlgorithm) ([]byte, error)
}

// SignatureAlgorithm represents a signature (or MAC) algorithm.
type SignatureAlgorithm string

// JSONWebKey represents a public or private key in JWK format.
type JSONWebKey struct {
	Key          interface{}
	Certificates []*x509.Certificate
	KeyID        string
	Algorithm    string
	Use          string
}
```

Reply:
Hah I spoke too soon:

```go
// OpaqueVerifier is an interface that supports verifying payloads with opaque
// public key(s). An OpaqueSigner may rotate signing keys transparently to the
// user of this interface.
type OpaqueVerifier interface {
	VerifyPayload(payload []byte, signature []byte, alg SignatureAlgorithm) error
}
```

I believe we can have `signer.Public().Key` return an `OpaqueVerifier` that allows us to have algorithm agility.

Reply:
Sounds good to me